crates/ethrpc/src/buffered.rs

```diff
@@ -293,12 +251,191 @@ where
     batches.for_each_concurrent(concurrency_limit, work)
 }
 
+/// Builds a metadata header for RPC requests.
+///
+/// This function takes an iterator of requests and their corresponding trace
+/// IDs, and generates a metadata string that groups the requests by their trace
+/// IDs and method names. The format of the output string is as follows:
+///
+/// `trace_id:method_name(index1,index2,...),method_name(index1,index2,...
+/// )|trace_id:...`
+///
+/// Each trace ID is followed by a colon and a list of method names. Each method
+/// name is followed by a list of indices (representing the position of the
+/// request in the original vector) enclosed in parentheses. Different method
+/// names are separated by commas. If there are multiple trace IDs, their
+/// entries are separated by a pipe character.
+///
+/// If a trace ID is `None`, it is represented as "null" in the output string.
+/// All requests with absent trace IDs are grouped together under "null".
+///
+/// # Arguments
+///
+/// * `requests` - A vector of tuples, where each tuple contains a request ID
+///   and a `Call` object representing the RPC request.
+/// * `trace_ids` - A vector of optional strings representing the trace IDs of
+///   the requests. The trace IDs correspond to the requests in the same
+///   position in the `requests` vector.
+///
+/// # Returns
+///
+/// This function returns a string representing the metadata header.
+///
+/// # Examples
+///
+/// ```
+/// let requests = vec![
+///     (
+///         1001,
+///         Call::MethodCall(MethodCall {
+///             jsonrpc: "2.0",
+///             method: "eth_call",
+///             params: vec![],
+///             id: Id::Num(1),
+///         }),
+///     ),
+///     (
+///         1001,
+///         Call::MethodCall(MethodCall {
+///             jsonrpc: "2.0",
+///             method: "eth_sendTransaction",
+///             params: vec![],
+///             id: Id::Num(2),
+///         }),
+///     ),
+///     (
+///         1002,
+///         Call::MethodCall(MethodCall {
+///             jsonrpc: "2.0",
+///             method: "eth_call",
+///             params: vec![],
+///             id: Id::Num(3),
+///         }),
+///     ),
+/// ];
+/// let trace_ids = vec![Some("1001".to_string()), None, Some("1002".to_string())];
+/// let metadata_header = build_rpc_metadata_header(requests, trace_ids);
+/// assert_eq!(
+///     metadata_header,
+///     "null:eth_sendTransaction(1)|1001:eth_call(0)|1002:eth_call(2)"
+/// );
+/// ```
+fn build_rpc_metadata_header<'a>(
+    requests: impl Iterator<Item = &'a (RequestId, Call)>,
+    trace_ids: impl Iterator<Item = &'a Option<String>>,
+) -> String {
+    let mut grouped_metadata: BTreeMap<String, BTreeMap<String, BTreeSet<usize>>> = BTreeMap::new();
+    for (idx, ((_, call), trace_id)) in requests.zip(trace_ids).enumerate() {
```

Comment:
I am totally lost, please add comments :pray:

Reply:
There is a doc explaining the format.

Reply:
Added more comments.